src/tree_builder/mod.rs

```diff
@@ -168,6 +171,72 @@ impl<Handle, Sink> TreeBuilder<Handle, Sink>
             frameset_ok: true,
             ignore_lf: false,
             foster_parenting: false,
+            context_elem: None,
+        }
+    }
+
+    /// Create a new tree builder which sends tree modifications to a particular `TreeSink`.
+    /// This is for parsing fragments.
+    ///
+    /// The tree builder is also a `TokenSink`.
+    pub fn new_for_fragment(mut sink: Sink,
+                            context_elem: Handle,
+                            form_elem: Option<Handle>,
+                            opts: TreeBuilderOpts) -> TreeBuilder<Handle, Sink> {
+        let doc_handle = sink.get_document();
+        let context_is_template =
+            sink.elem_name(context_elem.clone()) == QualName::new(ns!(HTML), atom!(template));
+        let mut tb = TreeBuilder {
+            opts: opts,
+            sink: sink,
+            mode: Initial,
+            orig_mode: None,
+            template_modes: if context_is_template { vec![InTemplate] } else { vec![] },
+            pending_table_text: vec!(),
+            quirks_mode: NoQuirks, // XXX set this to match the sink's document
+            doc_handle: doc_handle,
+            open_elems: vec!(),
+            active_formatting: vec!(),
+            head_elem: None,
+            form_elem: form_elem,
+            next_tokenizer_state: None,
+            frameset_ok: true,
+            ignore_lf: false,
+            foster_parenting: false,
+            context_elem: Some(context_elem),
+        };
+
+        // https://html.spec.whatwg.org/multipage/syntax.html#parsing-html-fragments
+        // 5. Let root be a new html element with no attributes.
+        // 6. Append the element root to the Document node created above.
+        // 7. Set up the parser's stack of open elements so that it contains just the single element root.
+        tb.create_root(vec!());
+        // 10. Reset the parser's insertion mode appropriately.
+        tb.mode = tb.reset_insertion_mode();
+
+        tb
+    }
+
+    // https://html.spec.whatwg.org/multipage/syntax.html#concept-frag-parse-context
+    // Step 4. Set the state of the HTML parser's tokenization stage as follows:
+    pub fn tokenizer_state_for_context_elem(&self) -> State {
+        let elem = self.context_elem.clone().expect("no context element");
```

Comment:
I'd prefer to handle non-HTML elements with an early return, to avoid rightward drift and for consistency with other code.